ergo-rest/src/api/peer_discovery_internals.rs

```diff
@@ -0,0 +1,410 @@
+                        timeout: Some(request_timeout_duration),
+                    };
+
+                    // If active, look up its peers.
+                    if get_info(node_conf).await.is_ok() {
+                        match get_peers_all(node_conf).await {
+                            Ok(peers) => {
+                                // It's important to send this message before the `AddActiveNode`
+                                // message below, to ensure an accurate `count` variable in task 1;
+                                // see (*) above in `peer_discovery_inner`.
+                                tx_peer.infallible_send(Msg::CheckPeers(peers)).await;
+                                tx_peer
+                                    .infallible_send(Msg::AddActiveNode(url.clone()))
+                                    .await;
+                            }
+                            Err(_) => {
+                                #[allow(clippy::unwrap_used)]
+                                tx_peer.infallible_send(Msg::AddInactiveNode(url)).await;
+                            }
+                        }
+                    } else {
+                        #[allow(clippy::unwrap_used)]
+                        tx_peer.infallible_send(Msg::AddInactiveNode(url)).await;
+                    }
+                });
+            }
+        })
+        .buffer_unordered(max_parallel_requests.get() as usize); // Allow for parallel requests
+
+    // Note: We need to define another binding to the spawn function to get around the Rust type
+    // checker.
+    #[cfg(not(target_arch = "wasm32"))]
+    let spawn_fn_new = tokio::spawn;
+
+    #[cfg(target_arch = "wasm32")]
+    let spawn_fn_new = wasm_bindgen_futures::spawn_local;
+
+    // (*) Run stream to completion.
+    spawn_fn_new(mapped_stream.for_each(|_| async move {}));
+}
+
+/// GET on /peers/all endpoint
+async fn get_peers_all(node: NodeConf) -> Result<Vec<PeerInfo>, NodeError> {
+    #[allow(clippy::unwrap_used)]
+    let url = node.addr.as_http_url().join("peers/all").unwrap();
+    let client = build_client(&node)?;
+    let rb = client.get(url);
+    let response = set_req_headers(rb, node).send().await?;
+    Ok(response.json::<Vec<PeerInfo>>().await?)
+}
+
+/// Used in the implementation of `peer_discovery`
+#[derive(Debug)]
+pub(crate) enum Msg {
+    /// Indicates that the ergo node at the given URL is active. This means that a GET request
+    /// to the node's /info endpoint responds with code 200 OK.
+    AddActiveNode(Url),
+    /// Indicates that the ergo node at the given URL is inactive. This means that a GET request
+    /// to the node's /info endpoint does not respond with code 200 OK.
+    AddInactiveNode(Url),
+    /// A list of peers of an active ergo node, returned from a GET on the /peers/all endpoint.
+    CheckPeers(Vec<PeerInfo>),
+}
+
+/// This trait abstracts over the `send` method of channel senders, assuming no failure.
```

Comment:
There are 2 types of channels to abstract over: `tokio::sync::mpsc::...` and `futures::channel::mpsc...`, hence these traits.